src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -181,6 +180,22 @@ public Task WriteStreamSuffixAsync()
             }
         }
 
+        public Task WriteRstStreamAsync(bool endReceived, Http2ErrorCode error)
+        {
+            lock (_dataWriterLock)
+            {
+                // Only reset if the connection is open or in a half closed state.
+                if (endReceived && _completed)
```

Comment:
I don't like evaluating endReceived in the lock. It really implies that Http2Stream.EndStreamReceived is protected by this lock, but it's not.

Reply:
@Tratcher pointed out this is an `&&` not `||` offline, so now I see why it was done this way. We decided to send the RST regardless since that's allowed by spec. The server/app shouldn't be likely to try to abort a dead stream anyway.

Reply:
Just add a comment saying that it was intentional that we don't take the lock or check completed in this method.